File: code/C#/Typeclasses/Concept/Algorithm.cs
using System.Collections.Generic;
using System.Linq;

namespace Typeclasses.Concept
{
    public static class Algorithm
    {
        public static List<T> Sort<T>(List<T> source, Ordered<T> model)
        {
            var result = new List<T>();
            foreach (var s in source)
            {
                var position = result.FindIndex(r => model.gt(r, s)) switch
                {
                    -1 => result.Count,
                    int p => p
                };
                result.Insert(position, s);
            }

            return result;
        }

        public static bool Equality<T>(List<T> list1, List<T> list2, Ordered<T> model) =>
            list1.Count == list2.Count && list1.Zip(list2).All(p => model.eq(p.First, p.Second));
    }
}

File: code/Scala/src/main/scala/net/sigusr/implicit0/Algorithm.scala
package net.sigusr.implicit0

import scala.collection.mutable.ArrayBuffer

object Algorithm {
  def Sort[T](source: Vector[T])(implicit model: Ordered[T]): Vector[T] = {
    val result = ArrayBuffer[T]()
    for (s <- source) {
      val position = result.indexWhere(r => model.gt(r, s)) match {
        case -1 => result.length
        case p => p
      }
      result.insert(position, s)
    }
    result.toVector
  }

  def Equality[T](vector1 : Vector[T], vector2 : Vector[T])(implicit model : Ordered[T]): Boolean =
    vector1.length == vector2.length && 
      vector1.zip(vector2).forall(p => model.eq(p._1, p._2))
}
